src/jit/lsra.cpp

```diff
@@ -4857,15 +4901,40 @@ LinearScan::tryAllocateFreeReg(Interval *currentInterval, RefPosition *refPositi
 //                  and that can be spilled.
 //
 // Arguments:
-//    current:         The interval for the current allocation
-//    refPosition:     The RefPosition of the current Interval for which a register is being allocated
+//    current               The interval for the current allocation
+//    refPosition           The RefPosition of the current Interval for which a register is being allocated
+//    allocationOptional    If true, a reg may not be allocated if all other ref positions currently
+//                          occupying registers are more important than the 'refPosition'.
 //
 // Return Value:
 //    The regNumber allocated to the RefPositon.  Returns REG_NA if no free register is found.
-
+//
+// Note:  Currently this routine uses weight and farthest distance of next reference
+// to select a ref position for spilling.  
+// a) if allocationOptional = false
+//        The ref position chosen for spilling will be the lowest weight
+//        of all and if there is is more than one ref position with the
+//        same lowest weight, among them choses the one with farthest
+//        distance to its next reference.
+// 
+// b) if allocationOptional = true
+//        The ref position chosen for spilling will not only be lowest weight
+//        of all but also has a weight lower than 'refPosition'.  If there is
+//        no such ref position, reg will not be allocated.
 regNumber
-LinearScan::allocateBusyReg(Interval *current, RefPosition *refPosition)
+LinearScan::allocateBusyReg(Interval* current,
+                            RefPosition* refPosition,
+                            bool allocationOptional)
 {
```

Comment:
First, the terms Requires and Optional should never be used together, it doesn't make any sense.
Second, I don't think it is the case that only a use can be optional - if I have a = a + b, then the def on the lhs could also be considered optional. That would require additional logic, I presume, but it is not fundamental.
Finally, other types of `RefPosition`s are also allocated - we call `tryAllocateReg` on these `RefType`s, but we do not attempt to spill if no register is found.
IMO, you should use RequiresRegister to mean places where a register is non-optional, then perhaps add something like `IsActualRef` to mean a reference on a node (`RefTypeDef` or `RefTypeUse`), where you might want to give it a "second chance". However, it may be the case that it would similarly be valuable to do a weight comparison, e.g. in the `RefTypeExpUse` case and spill something to load it if there are no registers available. So probably the right answer is to have a method, e.g. `SpillIfProfitable` that could be used to determine when to consider spilling for a `RefPosition` on which `RequiresRegister()` returns false.